dashboards/victoriametrics.json

```diff
@@ -1295,7 +1295,7 @@
           "steppedLine": false,
           "targets": [
             {
-              "expr": "vm_free_disk_space_bytes{job=\"$job\", instance=\"$instance\"} / (sum(rate(vm_rows_added_to_storage_total{job=\"$job\", instance=\"$instance\"}[1d])) * (sum(vm_data_size_bytes{job=\"$job\", instance=\"$instance\", type!=\"indexdb\"}) / sum(vm_rows{job=\"$job\", instance=\"$instance\", type!=\"indexdb\"})))",
+              "expr": "vm_free_disk_space_bytes{job=\"$job\", instance=\"$instance\"} / ignoring(path) ((sum(rate(vm_rows_added_to_storage_total{job=\"$job\", instance=\"$instance\"}[1d])) - sum without(type) (rate(vm_deduplicated_samples_total{job=\"$job\", instance=\"$instance\"}[1d]))) * (sum(vm_data_size_bytes{job=\"$job\", instance=\"$instance\", type!=\"indexdb\"}) / sum(vm_rows{job=\"$job\", instance=\"$instance\", type!=\"indexdb\"})))",
```

Comment:
Wow, QL becomes hard to read) Can we structure it a bit like the following:
```
vm_free_disk_space_bytes{job="$job", instance="$instance"} / 
ignoring(path) (
 (
  sum(rate(vm_rows_added_to_storage_total{job="$job", instance="$instance"}[1d])) - 
  sum(rate(vm_deduplicated_samples_total{job="$job", instance="$instance"}[1d])) without(type) 
 ) * 
 (
  sum(vm_data_size_bytes{job="$job", instance="$instance", type!="indexdb"}) /
  sum(vm_rows{job="$job", instance="$instance", type!="indexdb"})
 )
)
```

Such a shame we can't use `WITH` here... It seems to me query will be much easier to understand then:
```
WITH (
    filter = {job="$job", instance="$instance"},
    free_space = sum(vm_free_disk_space_bytes{filter}),
    samples_added:rate:1d = sum(rate(vm_rows_added_to_storage_total{filter}[1d])) - sum(rate(vm_deduplicated_samples_total{filter}[1d])) without(type),
    bytes_per_sample = sum(vm_data_size_bytes{filter, type!="indexdb"}) / sum(vm_rows{filter, type!="indexdb"})
)
free_space / (samples_added:rate:1d * bytes_per_sample)
```

I'm also a bit confused with `sum(rate(vm_deduplicated_samples_total{filter}[1d])) without(type)` - shouldn't it be `type=merge` only?

Reply:
> Wow, query becomes hard to read) Can we structure it a bit like the following:

I agree that the readability is poor.  JSON does not support multi-line strings however, so I can't think of a way to apply better structure.

Something to consider for the future:  a separate text file to define complex dashboard queries, that is combined with a json template to generate the final dashboard json.  Then:
  * we can format and document the queries nicely
  * perhaps `WITH` can be used, expanding it into a Prometheus-compatible query

> I'm also a bit confused with `sum(rate(vm_deduplicated_samples_total{filter}[1d])) without(type)` - shouldn't it be `type=merge` only?

You are right.  Thank you for raising it, this part I was actually not sure about.

So now from my understanding `vm_deduplicated_samples_total` has "merge" and "select" types:
  * `merge` - reflects removing duplicate points from storage
  * `select` - reflects dedup applied during a query because the storage dedup has not happened yet.   It's not related to storage size, so don't include it in the estimate.

Correcting the query will simplify it a little.

Reply:
> I agree that the readability is poor. JSON does not support multi-line strings however, so I can't think of a way to apply better structure.

I meant to make it more structured in Grafana query editor:
![image](https://user-images.githubusercontent.com/2902918/100288476-a836be00-2f6e-11eb-9480-ecb2f5f755e6.png)

^ This is supported by having `/n` chars in the query.

Reply:
> This is supported by having /n chars in the query.

I see.  It's a tradeoff because the "\n" chars will make the source less readable, but let's consider it again next time.